tempesta_fw/sock.c

```diff
@@ -78,6 +85,63 @@ skb_sender_cpu_clear(struct sk_buff *skb)
 #endif
 }
 
+/**
+ * Enters critical section synchronized with ss_synchronize().
+ * Active networking operations which involves SS callback calls must be
+ * protected by the guard: don't enter the section if the system is about
+ * to shutdown. The only exception is closing activity - this is the only
+ * activity allowed in progress of shutdown process.
+ *
+ * Returns zero (SS_OK) if we're in critical section and SS_BAD if shutdown
+ * process in progress and we can't enter the section.
+ */
+static int
+__active_guard_enter(unsigned long val, int cpu)
+{
+	atomic64_t *active = per_cpu_ptr(&__ss_active, cpu);
+	unsigned long a_new, a_old = atomic64_read(active);
+
+	while (1) {
+		if (unlikely(a_old >= SS_STOP_VAL))
+			return SS_BAD;
+
+		a_new = a_old + val;
+		a_new = atomic64_cmpxchg(active, a_old, a_new);
+		if (likely(a_new == a_old))
+			return SS_OK;
+		a_old = a_new;
+	}
+}
+
+static void
+__active_guard_exit(unsigned long val, int cpu)
+{
+	atomic64_sub(val, per_cpu_ptr(&__ss_active, cpu));
+}
+
+#define ss_active_guard_enter(v)	__active_guard_enter(v,		\
+							     smp_processor_id())
+#define ss_active_guard_exit(v)		__active_guard_exit(v,		\
+							    smp_processor_id())
+
+/**
+ * Guard for calling connection error/drop callback for each established socket,
+ * so we guarantee that all upper layer connections are closed.
+ */
+#define SS_CALL_GUARD_ENTER(cb, sk)					\
+({									\
+	__active_guard_enter(SS_SOFTIRQ_VAL, sk->sk_incoming_cpu);	\
+	SS_CALL(cb, sk);						\
+})
+
+#define SS_CALL_GUARD_EXIT(cb, sk)					\
+do {									\
+	if (unlikely(sk->sk_incoming_cpu == TFW_SK_CPU_INIT))		\
+		break;							\
+	SS_CALL(cb, sk);						\
+	__active_guard_exit(SS_SOFTIRQ_VAL, sk->sk_incoming_cpu);	\
+} while (0)
```

Comment:
The answer is "simplicity". All the macros are just several lines of code. Meantime, if a function is used, then the function must work with variadic arguments as well as accept address of the function, i.e. you have to pass full `((SsProto *)(sk)->sk_user_data)->hooks->callback_name` to each call of the function, instead of just pass name of the callback. At least I din't figure out easy and nice way to program it using functions. Frankly speaking I also don't see an issue with the macro...